Ruby implementation

luke-hill opened the pull request:
### 🤔 What's changed?

Just enough cucumber-query to satisfy the rerun formatter.

### ⚡️ What's your motivation? 

Move code that was written inside Cucumber-ruby over to this library. Get a first release out that we can build on.

### 🏷️ What kind of change is this?

<!--- Delete any options that are not relevant -->

- :bank: Refactoring/debt/DX (improvement to code design, tooling, etc. without changing behaviour)

### ♻️ Anything particular you want feedback on?

Are we following standard patterns? Is this a good foundation to build on?

### Checklist

- [x] Configure RubyGems trusted publishing for cucumber-query
- [ ] Do the release

luke-hill committed:
Initial files

luke-hill committed:
Add hidden files

luke-hill committed:
Fix typo

luke-hill committed:
Add cucumber messages to runtime requirements and port over existing queries

luke-hill committed:
Add in spec files

luke-hill committed:
Fix up all loading, relations and requirements

luke-hill committed:
Alter error class for new API

luke-hill committed:
Alter all other tests to pass

luke-hill committed:
Auto-gen rubocop config;

luke-hill committed:
Add note about duplication of code

luke-hill committed:
Minor linting and starting to unpick redundant pieces of spec support

luke-hill committed:
Lint / DRY up tests

mpkorstanje reviewed (commented):
It looks like you're writing tests for each method. That isn't necessary.

Rather you can define a list of `sources`:

* "../testdata/attachments.ndjson"
* "../testdata/empty.ndjson"
* "../testdata/hooks.ndjson"
* "../testdata/minimal.ndjson"
* "../testdata/rules.ndjson"
* "../testdata/examples-tables.ndjson"

And a map of query functions `queries` of the form `(Query query) ->  /* execute some query and return the results*/`. For example:

```
queries.put("findAllTestStepsStarted", (query) -> query.findAllTestStepsStarted().size());
```

Then take the product of `sources` and `queries` and for each of those populate a instance of the `Query` object with the messages from the source, and then apply the query function to the `Query` object.

Then you compare the result of query function to the `../testdata/<source>.<methodname>.results.json` file.

The nice thing is that you can add methods to the `queries` list as you implement them.

mpkorstanje commented on ruby/cucumber-query.gemspec:
```diff
@@ -0,0 +1,35 @@
+
+  s.add_dependency 'cucumber-messages', '> 25', '< 30'
+
+  s.add_development_dependency 'cucumber', '~> 10.1'
```
?

mpkorstanje commented on ruby/cucumber-query.gemspec:
```diff
@@ -0,0 +1,35 @@
+    'bug_tracker_uri' => 'https://github.com/cucumber/query/issues',
+    'changelog_uri' => 'https://github.com/cucumber/query/blob/main/CHANGELOG.md',
+    'documentation_uri' => 'https://github.com/cucumber/query/blob/main/CONTRIBUTING.md',
+    'mailing_list_uri' => 'https://groups.google.com/forum/#!forum/cukes',
```
?

mpkorstanje commented on ruby/cucumber-query.gemspec:
```diff
@@ -0,0 +1,35 @@
+  s.metadata = {
+    'bug_tracker_uri' => 'https://github.com/cucumber/query/issues',
+    'changelog_uri' => 'https://github.com/cucumber/query/blob/main/CHANGELOG.md',
+    'documentation_uri' => 'https://github.com/cucumber/query/blob/main/CONTRIBUTING.md',
```
Should be the read me I guess, not the contributors guide.

luke-hill committed:
Linting for tests and 1 bug fix where test case started was firing prematurely

mpkorstanje commented on ruby/lib/cucumber/query/hook_by_test_step.rb:
```diff
@@ -0,0 +1,32 @@
+      def hook_id(test_step)
+        return @hook_id_by_test_step_id[test_step.id] if @hook_id_by_test_step_id.key?(test_step.id)
+
+        raise TestStepUnknownError, "No hook found for #{test_step.id} }. Known: #{@hook_id_by_test_step_id.keys}"
```
The design of the query object assumes that if something can't be found either an empty list, empty optional or null is returned. There are no methods that throw.

This is also why all the methods are name `findX` or `findAllX` or findZByX`.

mpkorstanje reviewed (commented):
Some other nitpicks below, but I suppose those are there because this is WIP

mpkorstanje commented:
Removed WIP from title, marked PR as draft.

luke-hill committed:
Add in legacy ns;

luke-hill committed:
Ported over helpers from JS

luke-hill committed:
Initial implementation of initializer and update method

luke-hill committed:
Remove un-needed placeholder

luke-hill committed:
Regen the rubocop config;

luke-hill committed:
Remove all legacy query entities

luke-hill committed:
Move query into correct NS and rename helpers

luke-hill committed:
Remove legacy tests

luke-hill committed:
Fix readme

luke-hill committed:
Remove un-necessary items from package

mattwynne committed:
Move Ruby query implementation into gem

mattwynne committed:
Fix Ruby query packaging and test step lookup

mattwynne committed:
Add Ruby CI and release workflow

mattwynne committed:
Align Ruby gem version

mattwynne committed:
Update Ruby lockfile version

mattwynne committed:
Update changelog for Ruby implementation

mattwynne committed:
Pin Ruby workflow actions

mattwynne committed:
Disable bundler cache in Ruby release workflow

mattwynne committed:
Add RubyGems badge

mattwynne committed:
Tidy Ruby query implementation

mattwynne committed:
Merge branch 'main' into feature/ruby_implementation

luke-hill commented on .github/workflows/test-ruby.yml:
```diff
@@ -0,0 +1,56 @@
+      matrix:
+        os:
+          - ubuntu-latest
+        ruby: ['3.2', '3.3', '3.4', '3.5']
```
```suggestion
        ruby: ['3.2', '3.3', '3.4', '4.0']
```

luke-hill commented on .github/workflows/test-ruby.yml:
```diff
@@ -0,0 +1,56 @@
+        ruby: ['3.2', '3.3', '3.4', '3.5']
+        include:
+          - os: macos-latest
+            ruby: '3.4'
```
4.0 here and for all single ruby variants

luke-hill commented on ruby/Gemfile.lock:
```diff
@@ -0,0 +1,75 @@
+PATH
```
This file needs gitignoring

mpkorstanje replied:
@luke-hill why? To get consistent builds in CI, shouldn't we pin the versions?

luke-hill replied:
So we never track the gemfile.lock in gems because you will have diff versions for diff file systems. Plus we want to be able to have diff versions locally. All we should be ensuring is that we have valid versions.

For ci we should use lowspec gemfiles if we want to test diff version configurations using the build matrix. We do this in [cucumber-rails](https://github.com/cucumber/cucumber-rails/blob/main/.github/workflows/test-ruby.yml#L33)

luke-hill commented on ruby/VERSION:
```diff
@@ -0,0 +1 @@
+15.0.1
```
This numbers seems wrong. But I believe this shouldn't matter @mpkorstanje as it'll be overwritten on the next polyglot-release we call?

mpkorstanje replied:
Yes. Will be overwritten, but it would still be nice if it was correct to make the git history less confusing.

luke-hill reviewed (commented):
12/18 files marked as good. Few minor comments

luke-hill commented on ruby/lib/cucumber/query.rb:
```diff
@@ -0,0 +1,260 @@
+
+    # This method will be called with 1 of these 2 messages
+    #   [TestCaseStarted || TestCaseFinished]
+    def find_most_severe_test_step_result_by(message)
```
@mpkorstanje - Method 1 to review as/when you have time

luke-hill commented on ruby/lib/cucumber/query.rb:
```diff
@@ -0,0 +1,260 @@
+
+    # This method will be called with 1 of these 5 messages
+    #   [TestCase || TestCaseStarted || TestCaseFinished || TestStepStarted || TestStepFinished]
+    def find_pickle_by(message)
```
@mpkorstanje - Method 2 to review as/when you have time

luke-hill commented on ruby/lib/cucumber/repository.rb:
```diff
@@ -0,0 +1,147 @@
+      end
+    end
+
+    def update_gherkin_document(gherkin_document)
```
@mpkorstanje this is the only part of the repository update I think you need to review.

We previously spoke of Lineages but I think this is now considered good.

luke-hill commented on ruby/lib/cucumber/repository.rb:
```diff
@@ -0,0 +1,147 @@
+    }.freeze
+    private_constant :UPDATE_HANDLERS
+
+    def update(envelope)
```
Note to self / @mattwynne 

I'm proposing to write a small amount of custom code inside the `Envelope` message that'll return a type. Which should then make this a bit simpler/less metaprogrammy

luke-hill replied:
re-ping @mattwynne this is using a not yet released version. But I think you would agree it's much easier/simpler to read

luke-hill commented on ruby/lib/cucumber/query.rb:
```diff
@@ -0,0 +1,260 @@
+      repository.test_case_finished_by_test_case_started_id[test_case_started.id]
+    end
+
+    def find_test_run_duration
```
@mpkorstanje - This one I think can be ignored for review, but I had it flagged. Would be good to know if we have some edge case specs for this where the missing values are there.

luke-hill reviewed (commented):
Items previously mentioned for review @mpkorstanje

mattwynne committed:
Update .github/workflows/test-ruby.yml

Co-authored-by: Luke Hill <20105237+luke-hill@users.noreply.github.com>

mpkorstanje commented:
It's 30 degrees. To hot to look at this right now.

luke-hill committed:
Bump gemspec to only use ruby 3.3+ and higher dev gems

luke-hill committed:
Target Ruby 3.3+

luke-hill committed:
Fix CI

luke-hill committed:
Convert into new testing format for new query;

luke-hill committed:
Tidy up test iteration code to ensure we only select files that will run (let alone pass!)

luke-hill committed:
Add in more cck examples to be tested against

luke-hill committed:
Remove gemfile.lock

mpkorstanje committed:
Nitpicks

mpkorstanje commented on .github/workflows/test-ruby.yml:
```diff
@@ -0,0 +1,50 @@
+          ruby-version: ${{ matrix.ruby }}
+          bundler-cache: true
+          working-directory: ruby
+      - run: bundle exec rubocop --format simple
```
We don't do this in other projects.

mpkorstanje commented on .github/workflows/test-ruby.yml:
```diff
@@ -0,0 +1,50 @@
+          working-directory: ruby
+      - run: bundle exec rubocop --format simple
+        working-directory: ruby
+      - run: bundle exec rspec --format progress
```
I don't recognize this. In other projects we don't add `--format progress`.

mpkorstanje commented on ruby/lib/cucumber/query.rb:
```diff
@@ -0,0 +1,260 @@
+# Given one Cucumber Message, find another.
+#
+# Queries can be made while the test run is incomplete - and this will naturally return incomplete results
+# see <a href="https://github.com/cucumber/messages?tab=readme-ov-file#message-overview">Cucumber Messages - Message Overview</a>
```
I don't think html tags are valid in Ruby doc.

https://github.com/ruby/rdoc/blob/master/doc/markup_reference/rdoc.rdoc#multi-word-text-link

mpkorstanje commented on ruby/lib/cucumber/query.rb:
```diff
@@ -0,0 +1,260 @@
+      if message.is_a?(Cucumber::Messages::TestCaseStarted)
+        find_test_steps_finished_by(message)
+          .map(&:test_step_result)
+          .max_by { |test_step_result| test_step_result_rankings[test_step_result.status] }
```
Where is `test_step_result_rankings` defined?

luke-hill replied:
From line19: `include Cucumber::Messages::Helpers::TestStepResultComparator`

mpkorstanje commented on ruby/lib/cucumber/query.rb:
```diff
@@ -0,0 +1,260 @@
+        find_test_steps_finished_by(message)
+          .map(&:test_step_result)
+          .max_by { |test_step_result| test_step_result_rankings[test_step_result.status] }
+        # Java code: "PREVIOUS".max(comparing(TestStepResult::getStatus, new TestStepResultStatusComparator()));
```
Comments like this can be cleaned up.

mpkorstanje commented on ruby/lib/cucumber/query.rb:
```diff
@@ -0,0 +1,260 @@
+
+    def message_types
+      {
+        pickle_step: Cucumber::Messages::PickleStep,
```
Does Ruby have some sort of alias that lets you use the qualified name instead of the unqualified name? If so, this lookup table could be removed.

mpkorstanje reviewed (commented):
Looks okay. What does the code coverage look like?

luke-hill committed:
Fix version

luke-hill committed:
Merge branch 'main' into feature/ruby_implementation

luke-hill committed:
Fix for old ruby

luke-hill committed:
Fix requirements and rdoc

luke-hill committed:
Fix up test code

luke-hill committed:
Refactor out symbol refs to use verbose classes

mpkorstanje commented on ruby/lib/cucumber/repository.rb:
```diff
@@ -0,0 +1,147 @@
+      attachments_by_test_run_hook_started_id[attachment.test_run_hook_started_id] << attachment if attachment.test_run_hook_started_id
+    end
+
+    def update_feature(feature)
```
I would suggest removing this method and it's dependencies for now. You won't need it unless you implement the naming strategy which is a long way away. It's also not testable without implementing the naming strategy so you can safe on some work here.

luke-hill replied:
Update @mpkorstanje we need this method and "some" type of handler (Even if it's a nil handler), to handle the metaprogrammed enumeration.

We've added in 2 more "dumb" handlers for `undefined_parameter_type` and `source` as well as a catchall method that'll bubble up a warning that things aren't updating.

Merging this in AS IS, as it's essentially RtM - but there's no need to rush anything out for this, so it can just sit until someone builds a new version. 16.1+

mpkorstanje replied:
Then why not use a nil handler? Dead and untested code just slow you down.

luke-hill committed:
Fix required version of messages and improve handlers by utilising new message type on envelopes

luke-hill commented:
Note to viewers. CI is red whilst we move forward on messages being cut for a new major - Which would bring version parity in-line plus enable access to new methods

luke-hill committed:
Tidy up line length and remove comments

luke-hill committed:
Add in some more simple lambdas to validate more of the queries

luke-hill committed:
Fix findAll lambdas to return int

luke-hill committed:
Fix findMeta to return name of implementation

luke-hill committed:
Add in 2 new queries which also return simple counters;

luke-hill committed:
Bump version requirements for messages

luke-hill committed:
Merge branch 'main' into feature/ruby_implementation

luke-hill commented:
Updated with latest main and fixed version requirements to pick up a new enough version of messages to facilitate this working.

Should be green now

luke-hill committed:
Add in missing message handlers for 2 message updates plus catchall for new undefined ones
